### DISCUSSION TITLE
How to configure Go to use private module dependencies

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#12091] [Ideas]
### What is the issue?

ref: https://discord.com/channels/707636530424053791/1134415015198330981

We have a cookbook that explains how to clone private repositories here https://docs.dagger.io/710884/private-repositories/. Having said that, in some cases like go private modules a bit more tweaking is necessary. Here's an example of how I had to configure my dagger pipeline to make it work:

```go
   _, err = client.Pipeline("build").
        Container().
        From("golang:1.20").
        WithUnixSocket(sshAgentPath, client.Host().UnixSocket(sshAgentPath)).
        WithEnvVariable("SSH_AUTH_SOCK",sshAgentPath).
        WithDirectory("/app", hd).
    WithWorkdir("/app").
        WithExec([]string{"git", "config", "--global", `url.ssh://git@github.com/marcosnils/containers.insteadOf`, `https://github.com/marcosnils/containers`}).
        WithEnvVariable("GOPRIVATE", "github.com/marcosnils/containers").
        WithEnvVariable("GIT_SSH_COMMAND", "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=accept-new").
        WithExec([]string{"go", "build", "."}).
        Stdout(ctx)
    if err != nil {
        panic(err)
    }
```

It'd be nice to add a cookbook guide for this :pray: 

cc @vik

**Comment by TomChv** [comment]
@vikram-dagger I can help on this one, any thought? Since it's go specialised I don't know where we should write it

**Comment by vikram-dagger** [comment]
> @vikram-dagger I can help on this one, any thought? Since it's go specialised I don't know where we should write it

Thanks for contributing this. It can be written as a Go-specific guide, in the Guides area of the docs. We already have another one which is also Go-specific: https://docs.dagger.io/110632/embed-directories

**Comment by marcosnils** [comment]
:wave:  adding another option by using Githu's fine-grained tokens. ref: https://discord.com/channels/707636530424053791/1275877411497447466

```go
func (m *Lala) Test(src *dagger.Directory, token *dagger.Secret) *dagger.Container {
    return dag.
        Container().
        From("golang:1.23").
        WithDirectory("/app", src).
        WithWorkdir("/app").
        WithSecretVariable("GITHUB_TOKEN", token).
        WithExec([]string{"bash", "-c", `git config --global url."https://oauth2:${GITHUB_TOKEN}@github.com/marcosnils/containers".insteadOf https://github.com/marcosnils/containers`}).
        WithEnvVariable("GOPRIVATE", "github.com/marcosnils/containers").
        WithExec([]string{"go", "build", "."})
}
```

**Comment by jedevc** [comment]
@marcosnils is this separate from https://github.com/dagger/dagger/issues/8766?
